### PR TITLE
Bug/1995 fix loading state

### PR DESCRIPTION
**Pull request by eugene-manuilov**
## Summary

Addresses issue #1995

## Relevant technical choices

- Removed progress bar from `assets/js/modules/adsense/components/common/SiteSteps.js` because we don't need it there.

## Checklist

- [x] My code is tested and passes existing unit tests.
- [x] My code has an appropriate set of unit tests which all pass.
- [x] My code is backward-compatible with WordPress 4.7 and PHP 5.6.
- [x] My code follows the [WordPress](https://make.wordpress.org/core/handbook/best-practices/coding-standards/) coding standards.
- [x] My code has proper inline documentation.
- [x] I have added a QA Brief on the issue linked above.
- [x] I have signed the Contributor License Agreement (see <https://cla.developers.google.com/>).

### CODE REVIEW

File: assets/js/modules/adsense/components/settings/SettingsForm.js [hunk 1]
```diff
@@ -31,11 +31,17 @@ import {
 	ErrorNotice,
 	UseSnippetSwitch,
 } from '../common';
+import ProgressBar from '../../../../components/progress-bar';
 const { useSelect } = Data;
 
 export default function SettingsForm() {
 	const clientID = useSelect( ( select ) => select( STORE_NAME ).getClientID() );
 	const existingTag = useSelect( ( select ) => select( STORE_NAME ).getExistingTag() );
+	const hasResolvedGetExistingTag = useSelect( ( select ) => select( STORE_NAME ).hasFinishedResolution( 'getExistingTag' ) );
```
Comment (felixarntz) [inline assets/js/modules/adsense/components/settings/SettingsForm.js:40]: Nitpick, but it would probably make sense to combine these like you did above (since they belong to the same selector call in a way).

File: assets/js/modules/analytics/components/common/ProfileSelect.js [hunk 2]
```diff
@@ -51,7 +53,7 @@ export default function ProfileSelect() {
 		}
 	}, [ profileID ] );
 
-	if ( isLoadingAccounts || isLoadingProperties || isLoadingProfiles ) {
+	if ( ! hasResolvedAccounts || ( accountID && ! hasResolvedProperties ) || ! hasResolvedProfiles ) {
```
Comment (felixarntz) [inline assets/js/modules/analytics/components/common/ProfileSelect.js:56]: Why the `accountID` check in the second part of the if clause? I'd think if we add that, it would make sense to do the same for `accountID` and `propertyID` in the third segment since those are arguments to `getProfiles`.

File: assets/js/modules/analytics/components/common/ProfileSelect.test.js [hunk 3]
```diff
@@ -37,11 +37,20 @@ apiFetchMock.mockImplementation( ( ...args ) => {
...
-	dispatch( STORE_NAME ).setPropertyID( webPropertyId ); // eslint-disable-line sitekit/camelcase-acronyms
+	const {
+		id,
+		webPropertyId: propertyID, // eslint-disable-line sitekit/camelcase-acronyms
+		accountId: accountID, // eslint-disable-line sitekit/camelcase-acronyms
+	} = fixtures.propertiesProfiles.profiles[ 0 ];
+
+	dispatch( STORE_NAME ).finishResolution( 'getAccounts', [] );
+	dispatch( STORE_NAME ).finishResolution( 'getProperties', [ accountID ] );
+	dispatch( STORE_NAME ).finishResolution( 'getProperties', [ '0' ] );
```
Comment (felixarntz) [inline assets/js/modules/analytics/components/common/ProfileSelect.test.js:48]: Why is that here? Seems like there's something else we need to fix instead... Also, I think we should call `receiveGetAccounts` and `receiveGetProperties` here with applicable fixture values too. While they're technically not used in the component, not passing that data while at the same time marking the resolver resolution as finished causes a state that shouldn't exist.

File: assets/js/modules/analytics/components/common/ProfileSelect.test.js [hunk 4]
```diff
@@ -50,6 +59,10 @@ const setupRegistryWithExistingTag = ( { dispatch } ) => {
 		propertyID: fixtures.accountsPropertiesProfiles.profiles[ 0 ].webPropertyId, // eslint-disable-line sitekit/camelcase-acronyms
 	};
 	const { id } = fixtures.propertiesProfiles.profiles[ 0 ];
+
+	dispatch( STORE_NAME ).finishResolution( 'getAccounts', [] );
+	dispatch( STORE_NAME ).finishResolution( 'getProperties', [ existingTag.accountID ] );
+	dispatch( STORE_NAME ).finishResolution( 'getProfiles', [ existingTag.accountID, existingTag.propertyID ] );
```
Comment (felixarntz) [inline assets/js/modules/analytics/components/common/ProfileSelect.test.js:65]: See above, let's also populate data for these selectors.

File: assets/js/modules/analytics/components/common/ProfileSelect.test.js [hunk 5]
```diff
@@ -60,7 +73,13 @@ const setupRegistryWithExistingTag = ( { dispatch } ) => {
 const setupEmptyRegistry = ( { dispatch } ) => {
 	const accountID = fixtures.accountsPropertiesProfiles.profiles[ 0 ].accountId; // eslint-disable-line sitekit/camelcase-acronyms
 	const propertyID = fixtures.accountsPropertiesProfiles.profiles[ 0 ].webPropertyId; // eslint-disable-line sitekit/camelcase-acronyms
+
+	dispatch( STORE_NAME ).finishResolution( 'getAccounts', [] );
+	dispatch( STORE_NAME ).finishResolution( 'getProperties', [ accountID ] );
+	dispatch( STORE_NAME ).finishResolution( 'getProfiles', [ accountID, propertyID ] );
```
Comment (felixarntz) [inline assets/js/modules/analytics/components/common/ProfileSelect.test.js:79]: See above.

File: assets/js/modules/analytics/components/common/PropertySelect.js [hunk 6]
```diff
@@ -50,10 +50,10 @@ export default function PropertySelect() {
 		return data;
 	} );
 
-	const hasExistingTag = useSelect( ( select ) => select( STORE_NAME ).hasExistingTag() );
-	const hasGTMPropertyID = useSelect( ( select ) => !! select( MODULES_TAGMANAGER ).getSingleAnalyticsPropertyID() );
 	const propertyID = useSelect( ( select ) => select( STORE_NAME ).getPropertyID() );
 	const hasResolvedAccounts = useSelect( ( select ) => select( STORE_NAME ).hasFinishedResolution( 'getAccounts' ) );
+	const hasExistingTag = useSelect( ( select ) => select( STORE_NAME ).hasExistingTag() );
+	const hasGTMPropertyID = useSelect( ( select ) => !! select( MODULES_TAGMANAGER ).getSingleAnalyticsPropertyID() );
```
Comment (felixarntz) [inline assets/js/modules/analytics/components/common/PropertySelect.js:56]: Is there a specific reason you moved these? Just curious since I'd imagine this doesn't change anything?

File: assets/js/modules/tagmanager/components/common/AMPContainerSelect.test.js [hunk 7]
```diff
@@ -50,6 +50,8 @@ describe( 'AMPContainerSelect', () => {
 		registry.dispatch( STORE_NAME ).setAccountID( accountID );
 		registry.dispatch( STORE_NAME ).receiveGetAccounts( [ account ] );
 		registry.dispatch( STORE_NAME ).receiveGetContainers( [ ...webContainers, ...ampContainers ], { accountID } );
+		registry.dispatch( STORE_NAME ).finishResolution( 'getAccounts', [] );
+		registry.dispatch( STORE_NAME ).finishResolution( 'getContainers', [ accountID ] );
```
Comment (felixarntz) [inline assets/js/modules/tagmanager/components/common/AMPContainerSelect.test.js:54]: Maybe we can group each `finishResolution` call with the corresponding `receiveGet...` call? That would make the connection more clear when reading the code.

File: assets/js/modules/tagmanager/components/common/AMPContainerSelect.test.js [hunk 8]
```diff
@@ -70,6 +72,8 @@ describe( 'AMPContainerSelect', () => {
 		registry.dispatch( STORE_NAME ).setAccountID( accountID );
 		registry.dispatch( STORE_NAME ).receiveGetAccounts( [ account ] );
 		registry.dispatch( STORE_NAME ).receiveGetContainers( containers, { accountID } );
+		registry.dispatch( STORE_NAME ).finishResolution( 'getAccounts', [] );
+		registry.dispatch( STORE_NAME ).finishResolution( 'getContainers', [ accountID ] );
```
Comment (felixarntz) [inline assets/js/modules/tagmanager/components/common/AMPContainerSelect.test.js:76]: See above (same applies to the rest of this file).

File: assets/js/modules/tagmanager/components/common/AMPContainerSelect.test.js [hunk 9]
```diff
@@ -135,6 +147,8 @@ describe( 'AMPContainerSelect', () => {
 		const accountID = account.accountId; // eslint-disable-line sitekit/camelcase-acronyms
 		registry.dispatch( STORE_NAME ).receiveGetAccounts( [ account ] );
 		registry.dispatch( STORE_NAME ).setAccountID( accountID );
+		registry.dispatch( STORE_NAME ).finishResolution( 'getAccounts', [] );
+		registry.dispatch( STORE_NAME ).finishResolution( 'getContainers', [ accountID ] );
```
Comment (felixarntz) [inline assets/js/modules/tagmanager/components/common/AMPContainerSelect.test.js:151]: Why is this marked as finished here? I'd think this should explicitly _not_ be finished here, to trigger the loading state in the UI.

File: assets/js/modules/tagmanager/components/common/AMPContainerSelect.test.js [hunk 10]
```diff
@@ -149,6 +163,8 @@ describe( 'AMPContainerSelect', () => {
 		registry.dispatch( STORE_NAME ).receiveGetAccounts( [ account ] );
 		registry.dispatch( STORE_NAME ).receiveGetContainers( containers, { accountID } );
 		registry.dispatch( CORE_SITE ).receiveSiteInfo( { ampMode: AMP_MODE_PRIMARY } );
+		registry.dispatch( STORE_NAME ).finishResolution( 'getAccounts', [] );
+		registry.dispatch( STORE_NAME ).finishResolution( 'getContainers', [ accountID ] );
```
Comment (felixarntz) [inline assets/js/modules/tagmanager/components/common/AMPContainerSelect.test.js:167]: See above comment regarding grouping.

File: assets/js/modules/tagmanager/components/common/AMPContainerSelect.test.js [hunk 11]
```diff
@@ -162,6 +178,8 @@ describe( 'AMPContainerSelect', () => {
 		registry.dispatch( STORE_NAME ).receiveGetAccounts( [ account ] );
 		registry.dispatch( STORE_NAME ).receiveGetContainers( containers, { accountID } );
 		registry.dispatch( CORE_SITE ).receiveSiteInfo( { ampMode: AMP_MODE_SECONDARY } );
+		registry.dispatch( STORE_NAME ).finishResolution( 'getAccounts', [] );
+		registry.dispatch( STORE_NAME ).finishResolution( 'getContainers', [ accountID ] );
```
Comment (felixarntz) [inline assets/js/modules/tagmanager/components/common/AMPContainerSelect.test.js:182]: See above.

File: assets/js/modules/tagmanager/components/common/ContainerSelect.js [hunk 12]
```diff
@@ -43,12 +43,16 @@ export default function ContainerSelect( {
...
+		hasResolvedAccounts: select( STORE_NAME ).hasFinishedResolution( 'getAccounts' ),
+	} ) );
+
 	const accountID = useSelect( ( select ) => select( STORE_NAME ).getAccountID() );
 	const hasExistingTag = useSelect( ( select ) => select( STORE_NAME ).hasExistingTag() );
-	const isLoadingContainers = useSelect( ( select ) => select( STORE_NAME ).isDoingGetContainers( accountID ) );
+	const hasResolvedContainers = useSelect( ( select ) => select( STORE_NAME ).hasFinishedResolution( 'getContainers', [ accountID ] ) );
 
-	if ( accounts === undefined || containers === undefined || isLoadingContainers ) {
+	if ( accounts === undefined || ! hasResolvedAccounts || containers === undefined || ! hasResolvedContainers ) {
```
Comment (felixarntz) [inline assets/js/modules/tagmanager/components/common/ContainerSelect.js:55]: I think the `undefined` checks still need to be removed from here.

File: assets/js/modules/tagmanager/components/common/WebContainerSelect.test.js [hunk 13]
```diff
@@ -50,6 +50,8 @@ describe( 'WebContainerSelect', () => {
 		registry.dispatch( STORE_NAME ).setAccountID( accountID );
 		registry.dispatch( STORE_NAME ).receiveGetAccounts( [ account ] );
 		registry.dispatch( STORE_NAME ).receiveGetContainers( [ ...webContainers, ...ampContainers ], { accountID } );
+		registry.dispatch( STORE_NAME ).finishResolution( 'getAccounts', [] );
+		registry.dispatch( STORE_NAME ).finishResolution( 'getContainers', [ accountID ] );
```
Comment (felixarntz) [inline assets/js/modules/tagmanager/components/common/WebContainerSelect.test.js:54]: See above.

**Review by felixarntz — CHANGES_REQUESTED** [review]
This mostly looks solid, I mostly have a few follow-up questions and some minor suggestions.

File: assets/js/modules/adsense/components/settings/SettingsForm.js (hunk 1 above)
Comment (eugene-manuilov) [reply assets/js/modules/adsense/components/settings/SettingsForm.js]: Updated.

File: assets/js/modules/analytics/components/common/PropertySelect.js (hunk 6 above)
Comment (eugene-manuilov) [reply assets/js/modules/analytics/components/common/PropertySelect.js]: I don't remember why it has been moved. Reverted to its original state.

File: assets/js/modules/tagmanager/components/common/AMPContainerSelect.test.js (hunk 7 above)
Comment (eugene-manuilov) [reply assets/js/modules/tagmanager/components/common/AMPContainerSelect.test.js]: Yep, makes sense. Updated.

File: assets/js/modules/tagmanager/components/common/AMPContainerSelect.test.js (hunk 8 above)
Comment (eugene-manuilov) [reply assets/js/modules/tagmanager/components/common/AMPContainerSelect.test.js]: Updated.

File: assets/js/modules/tagmanager/components/common/AMPContainerSelect.test.js (hunk 9 above)
Comment (eugene-manuilov) [reply assets/js/modules/tagmanager/components/common/AMPContainerSelect.test.js]: Good catch! Thanks! Updated to fix this issue.

File: assets/js/modules/tagmanager/components/common/AMPContainerSelect.test.js (hunk 10 above)
Comment (eugene-manuilov) [reply assets/js/modules/tagmanager/components/common/AMPContainerSelect.test.js]: Updated.

File: assets/js/modules/tagmanager/components/common/AMPContainerSelect.test.js (hunk 11 above)
Comment (eugene-manuilov) [reply assets/js/modules/tagmanager/components/common/AMPContainerSelect.test.js]: Updated.

File: assets/js/modules/tagmanager/components/common/ContainerSelect.js (hunk 12 above)
Comment (eugene-manuilov) [reply assets/js/modules/tagmanager/components/common/ContainerSelect.js]: Yep, updated.

File: assets/js/modules/tagmanager/components/common/WebContainerSelect.test.js (hunk 13 above)
Comment (eugene-manuilov) [reply assets/js/modules/tagmanager/components/common/WebContainerSelect.test.js]: Updated.

File: assets/js/modules/analytics/components/common/ProfileSelect.js (hunk 2 above)
Comment (eugene-manuilov) [reply assets/js/modules/analytics/components/common/ProfileSelect.js]: Ok, looks like we don't need it anymore, but in the beginning, I had a situation when profiles select got loaded faster than properties select and having `accountID` check here helped to solve that race condition.

I have removed `accountID`, so it should be good now.

File: assets/js/modules/analytics/components/common/ProfileSelect.test.js (hunk 3 above)
Comment (eugene-manuilov) [reply assets/js/modules/analytics/components/common/ProfileSelect.test.js]: It is required for the `should be disabled when in the absence of an valid account or property ID.` test. I have updated that test to finish this resolution there and added fixture values too.

File: assets/js/modules/analytics/components/common/ProfileSelect.test.js (hunk 4 above)
Comment (eugene-manuilov) [reply assets/js/modules/analytics/components/common/ProfileSelect.test.js]: Updated.

File: assets/js/modules/analytics/components/common/ProfileSelect.test.js (hunk 5 above)
Comment (eugene-manuilov) [reply assets/js/modules/analytics/components/common/ProfileSelect.test.js]: Updated.

File: assets/js/modules/analytics/components/common/ProfileSelect.js (hunk 2 above)
Comment (eugene-manuilov) [reply assets/js/modules/analytics/components/common/ProfileSelect.js]: Actually, there was an issue when nothing was selected on the setup page. I have updated ProfileSelect and PropertySelect components to check whether fetching is started to prevent progress bars from appearing unexpectedly. Also added a new story to the Analytics setup storybook.

File: assets/js/modules/analytics/components/common/ProfileSelect.js [hunk 14]
```diff
@@ -34,13 +34,41 @@ import { trackEvent } from '../../../../util';
...
+		hasResolvedProfiles,
+	} = useSelect( ( select ) => {
+		const data = {
+			accountID: select( STORE_NAME ).getAccountID(),
+			propertyID: select( STORE_NAME ).getPropertyID(),
+			profiles: [],
+			hasStartedFetchingProperties: false,
+			hasResolvedProperties: false,
+			hasStartedFetchingProfiles: false,
+			hasResolvedProfiles: false,
```
Comment (felixarntz) [inline assets/js/modules/analytics/components/common/ProfileSelect.js:56]: In this case, wouldn't we be able to use the `isResolving` meta selector instead? Then we could have e.g. `isResolvingProperties` which I believe would do the same as `hasStartedFetchingProperties && ! hasResolverProperties`?

File: assets/js/modules/analytics/components/common/PropertySelect.js [hunk 15]
```diff
@@ -35,15 +35,22 @@ import { trackEvent } from '../../../../util';
...
+		accountID,
+		properties,
+		hasStartedFetchingProperties,
+		hasResolvedProperties,
+	} = useSelect( ( select ) => {
 		const data = {
 			accountID: select( STORE_NAME ).getAccountID(),
 			properties: [],
+			hasStartedFetchingProperties: false,
 			hasResolvedProperties: false,
```
Comment (felixarntz) [inline assets/js/modules/analytics/components/common/PropertySelect.js:48]: See above, we might be able to simplify this via `isResolving`.

File: assets/js/modules/analytics/components/common/PropertySelect.js [hunk 16]
```diff
@@ -64,7 +71,7 @@ export default function PropertySelect() {
 		}
 	}, [ propertyID ] );
 
-	if ( ! hasResolvedAccounts || ( accountID && ! hasResolvedProperties ) ) {
+	if ( ! hasResolvedAccounts || ( hasStartedFetchingProperties && ! hasResolvedProperties ) ) {
```
Comment (felixarntz) [inline assets/js/modules/analytics/components/common/PropertySelect.js:74]: Why is this extra check for whether resolving has already started necessary? If there's nothing selected, shouldn't the resolver still have to be resolved either way (it would be the resolver for e.g. a value of `undefined`).

**Review by felixarntz — COMMENTED** [review]
A few follow-up questions here, but this should be almost good to go.

File: assets/js/modules/analytics/components/common/ProfileSelect.js (hunk 14 above)
Comment (eugene-manuilov) [reply assets/js/modules/analytics/components/common/ProfileSelect.js]: Ah... yes, you are right. We can use `isResolving` here. Updated.

File: assets/js/modules/analytics/components/common/PropertySelect.js (hunk 15 above)
Comment (eugene-manuilov) [reply assets/js/modules/analytics/components/common/PropertySelect.js]: Updated to use `isResolving`.

File: assets/js/modules/analytics/components/common/PropertySelect.js (hunk 16 above)
Comment (eugene-manuilov) [reply assets/js/modules/analytics/components/common/PropertySelect.js]: I have updated it to use `isResolving` which is more accurate and strictly identifies when we are fetching something or not. It helps us to show the progress bar only when an actual process of fetching is happening.

**Review by felixarntz — APPROVED** [review]
Awesome!
